rpc: Optimize serialization and enhance metadata of dumptxoutset output

fjahr opened the pull request:
The second attempt at implementing the `dumptxoutset` space optimization as suggested in #25675. Closes #25675.

This builds on the work done in #26045, addresses open feedback, adds some further improvements (most importantly usage of compact size), documentation, and an additional test.

The [original snapshot at height 830,000](https://github.com/bitcoin/bitcoin/pull/29551) came in at 10.82 GB. With this change, the same snapshot is 8.94 GB, a reduction of 17.4%.

This also enhances the metadata of the output file and adds the following data to allow for better error handling and make future upgrades easier:
- A newly introduced utxo set magic
- A version number
- The network magic
- The block height

fjahr committed:
rpc: Optimize serialization disk space of dumptxoutset

Co-authored-by: Aurèle Oulès <aurele@oules.com>
Co-authored-by: TheCharlatan <seb.kung@gmail.com>

fjahr committed:
assumeutxo: Add test for changed coin size value

fjahr committed:
assumeutxo: Add documentation on dumptxoutset serialization format

Ferrydh commented:
这是来自QQ邮箱的假期自动回复邮件。
 
您好，我最近正在休假中，无法亲自回复您的邮件。我将在假期结束后，尽快给您回复。

Sjors commented:
> 我最近正在休假中

Well, I'm not on holiday :-)

tACK b0cfbce17a84158973ebba16f9801bd57cb70fe6

I do suggest adding a comment to `PopulateAndValidateSnapshot` in `validation.h`:

> To reduce space, this function takes advantage of the guarantee by leveldb that keys are lexicographically sorted.

I generated a snapshot and loaded it using the patch in #29551 (only waited for it to reach the tip, not a full background sync).

For those testing who don't want to wait out the long flush, see #28358.

fjahr commented:
Thanks for the review and testing!

> I do suggest adding a comment to `PopulateAndValidateSnapshot` in `validation.h`:
> 
> > To reduce space, this function takes advantage of the guarantee by leveldb that keys are lexicographically sorted.

As it is worded, wouldn't this be more appropriate to add in `CreateUTXOSnapshot`? That's where we are iterating with the pcursor, leveraging the sorting, and writing the coins to file, so I would say that's also where the space savings are realized. 

I have now amended the comment a bit and added a version of it to both `CreateUTXOSnapshot` and `PopulateAndValidateSnapshot`. Please let me know if you think this is alright: https://github.com/bitcoin/bitcoin/pull/29612/commits/97eb214ec3062ab07f3a340725b8c418b330b97b

sedited reviewed (approved):
Nice, ACK 923ec9043af668f3a342422d32c1c28d29018ed0

Just some comment nits that could easily be done in a follow-up too.

fjahr commented on test/functional/feature_assumeutxo.py:
```diff
@@ -98,12 +98,16 @@ def expected_error(log_msg="", rpc_details=""):
-            [b"\x80", 36, "091e893b3ccb4334378709578025356c8bcb0a623f37c7c4e493133c988648e5"],  # another wrong coin code
+            # The coin size byte is skipped here and its manipulation is tested
+            # below because this will always cause parsing to fail, resulting in
+            # and error that is not a mismatch of the expected hash.
```
yepp, `in an error` would be right

fjahr replied:
Do you think it is clear otherwise?

TheCharlatan replied:
Could be a bit clearer I guess, but I don't have a suggestion :P.

fjahr replied:
Fixed and tried a different wording to make it more clear

fjahr commented on src/rpc/blockchain.cpp:
```diff
@@ -2669,21 +2670,49 @@ UniValue CreateUTXOSnapshot(
     unsigned int iter{0};
+    std::vector<std::pair<uint32_t, Coin>> coins;
+
+    // To reduce space the serialization format of the snapshot avoids
```
Sounds good, will think about what that could look like if I have to retouch or in the follow-up.

fjahr replied:
I spent some time thinking about a diagram that would work but I wasn't very happy what I could come up with. I instead copied the expression used in the issue. It shows the format in a concise way that I find understandable: `list(Txid, list((vout,Coin))`. Let me know what you think.

TheCharlatan replied:
Maybe also add the size of the coins list `list(Txid, number_of_coins, list((vout,Coin)))`?

fjahr replied:
I will add this if I have to retouch, but if not I can do this as a follow-up in one of my other assumeutxo PRs.

fjahr commented on src/rpc/blockchain.cpp:
```diff
@@ -2669,21 +2670,49 @@ UniValue CreateUTXOSnapshot(
+    // leveldb that keys are lexicographically sorted.
+    // In the coins vector we collect all coins that belong to a certain tx hash
+    // (key.hash) and when we have them all (key.hash != last_hash) we write
+    // them to file using the below lamda function.
```
fixed

fjahr commented on src/rpc/blockchain.cpp:
```diff
@@ -2669,21 +2670,49 @@ UniValue CreateUTXOSnapshot(
+    std::vector<std::pair<uint32_t, Coin>> coins;
+
+    // To reduce space the serialization format of the snapshot avoids
+    // dublication of tx hashes. The code takes advantage of the guarantee by
```
fixed

sedited reviewed (approved):
Re-ACK 71d41d536bc2f47a014c898891d77d58502c9554

theStack reviewed (commented):
Concept ACK

mzumsande commented on src/rpc/blockchain.cpp:
```diff
@@ -2693,21 +2694,42 @@ UniValue CreateUTXOSnapshot(
+                last_hash = key.hash;
+                coins.clear();
+            }
+            coins.emplace_back(key.n, coin);
```
slightly unrelated nit: might be useful to assert that we actually wrote the right number of coins (we just take `maybe_stats->coins_count` and wouldn't abort if the actual number of coins written was different becaus the db changed for some reason). I'm sure https://github.com/bitcoin/bitcoin/pull/15606#discussion_r274479369 is correct, but checking it explicitly seems better.

fjahr replied:
I added a counter and check it at the end

mzumsande reviewed (commented):
Light ACK 71d41d536bc2f47a014c898891d77d58502c9554

I tested this on signet and it worked well, but I'm not very familiar with our serialization code.

theStack reviewed (approved):
Code-review ACK 71d41d536bc2f47a014c898891d77d58502c9554

(Fwiw, I rebased #27432 on this PR and adapted the utxo-to-sqlite3 tool to the new serialization format.)

Sjors commented:
re-utACK 71d41d536bc2f47a014c898891d77d58502c9554

Only test (rebase) and documentation changes since my last review.

fjahr commented:
Thanks all for the review, unless I have to retouch (I hope not since this has collected 4 ACKs) I will address the comments in a follow-up.

achow101 reviewed (commented):
`dumptxoutset` is a RPC that has been available for several releases, including those prior to the completion of assumeutxo validation code. Changing the format now is a breaking change - utxo sets dumped prior to the PR are incompatible with any version with this PR merged, and utxo sets dumped after this PR are incompatible with all prior versions.

Conceivably, in the future, someone could run an older node and create a utxo set snapshot to setup a new node with more recent software. This would then fail as the formats are incompatible. Currently the failure does not provide much detail on what actually went wrong. The RPC error is just "Unable to load UTXO snapshot" while the log does say "bad snapshot format or truncated snapshot after deserializing 0 coins". However that is not an indication of a version incompatibility, and I think it would be better if the new format could include a version number so that this could be properly reported.

Having a version number would also make it easy to distinguish which snapshot format to deserialize as so we can avoid these kinds of errors. We could detect which format was used for serialization and proceed to deserialize as necessary.

The reverse is also a possibility - a new node creates a snapshot with the new format and tries to load it into older software, but this seems unlikely to actually happen. But, it would also be trivial to add an option that allows for serializing in the old format for those that need it.

I think this also needs a release note about the format change as utxo set snapshots are useful for more than just snapshot loading.

mzumsande commented:
> utxo set snapshots are useful for more than just snapshot loading.

Do you know of any examples of the utxo dumps already being used for something else?

achow101 commented:
> Do you know of any examples of the utxo dumps already being used for something else?

For myself, I have a couple of random scripts that read utxo dumps in order to compute some statistics and other info on the utxo set. I'm not aware of any actual projects using utxo snapshots, but I also haven't actively looked.

Sjors commented:
I like magic bytes. For PSBT it's just "psbt" in ascii followed by 0xff: https://en.bitcoin.it/wiki/BIP_0174

So you could use "utxo" in ascii: 0x75 0x74 0x78 0x6F 0xFF

But then while you're at it, you could also append:
* ChainTypeToString ("main", "testnet", "signet", "regtest")
* height (so we can have an error like "you tried to load a snapshot at height X, but only heights Y, Z are supported by this version of PACKAGE_NAME. Newer releases may support more recent snapshots.")
* for signet: the signet challenge bytes

fjahr commented on src/rpc/blockchain.cpp:
```diff
@@ -2693,21 +2694,42 @@ UniValue CreateUTXOSnapshot(
 
+    std::map<uint256, std::vector<std::pair<uint32_t, Coin>>> mapCoins;
     COutPoint key;
+    uint256 last_hash;
```
done

fjahr commented on src/validation.cpp:
```diff
@@ -5564,69 +5564,81 @@ bool ChainstateManager::PopulateAndValidateSnapshot(
+            size_t size{0};
+            size = ReadCompactSize(coins_file);
+
+            if(size > coins_left) {
```
fixed

fjahr commented on src/validation.cpp:
```diff
@@ -5564,69 +5564,81 @@ bool ChainstateManager::PopulateAndValidateSnapshot(
+            Txid txid;
+            coins_file >> txid;
+            size_t size{0};
+            size = ReadCompactSize(coins_file);
```
renamed

fjahr commented on src/rpc/blockchain.cpp:
```diff
@@ -2693,21 +2694,42 @@ UniValue CreateUTXOSnapshot(
 
     afile << metadata;
 
+    std::map<uint256, std::vector<std::pair<uint32_t, Coin>>> mapCoins;
```
leftover from a previous approach,  removed

fjahr commented:
So far I have addressed the minor feedback in the latest push and I am now working on the extended metadata for the snapshot. I am currently planning to add the following:

1. magic bytes - the ones suggested by @Sjors 
2. snapshot version
3. Network magic (pchMessageStart) - This lets us distinguish between the different networks and since it’s generated from the challenge for signets, we can deduce that it’s probably a custom signet if we don’t recognize them and give an appropriate feedback to the user
4. block height

Let me know if you disagree with this choice or have something else to add, otherwise I will have the code ready soon.

Sjors commented:
@fjahr sounds good. Let me know when that's ready and I'll bake a halving block snapshot...

fjahr committed:
rpc: Enhance metadata of the dumptxoutset output

The following data is added:
- A newly introduced utxo set magic
- A version number
- The network magic
- The block height

fjahr commented:
Alright, ready for review again. I hope I have addressed all the ideas that were mentioned as intended. Sorry but the enhancement of the metadata is now one big commit for now, I can try to split it up if it makes review easier.

Sjors commented:
> I can try to split it up if it makes review easier.

Doesn't matter that much, but it's probably to put the `kernel/chainparams.h` change in its own commit so e.g. @TheCharlatan can do a drive-by review.

Here's some new snapshot torrents (not sure if I'm seeding them correctly...).

Torrent for testnet: `magnet:?xt=urn:btih:787f5917029876c83301c49dc97bb69224597285&dn=utxo-testnet-2500000.dat&tr=udp%3A%2F%2Ftracker.bitcoin.sprovoost.nl%3A6969`

Torrent for signet: `magnet:?xt=urn:btih:715a8797d6154a2474f225f6019364b6230eed17&dn=utxo-signet-160000.dat&tr=udp%3A%2F%2Ftracker.bitcoin.sprovoost.nl%3A6969`

Torrent for mainnet at the halving block (I'll update #28553 to use this): `magnet:?xt=urn:btih:55cb2ea9b6c48aa11d676de871f9639e3fa3dd7e&dn=utxo-840000.dat&tr=udp%3A%2F%2Ftracker.bitcoin.sprovoost.nl%3A6969`

TheCharlatan commented on src/kernel/chainparams.cpp:
```diff
@@ -542,3 +543,33 @@ std::unique_ptr<const CChainParams> CChainParams::TestNet()
+    const auto signet_msg = CChainParams::SigNet(signet_opts)->MessageStart();
+
+    if (std::equal(message.begin(), message.end(), mainnet_msg.data())) {
+        return "Mainnet";
```
I'm surprised we never had to match the magic bytes to a chaintype. Instead of defining new strings here though, could you use `ChainTypeToString(ChainType::MAIN)`?

fjahr replied:
Me too :) Yes, that makes sense. I opted for the function to return the `ChainType` and then I use `ChainTypeToString` on the result.

TheCharlatan commented on src/kernel/chainparams.h:
```diff
@@ -183,4 +185,6 @@ class CChainParams
     ChainTxData chainTxData;
 };
 
+std::optional<std::string> GetNetworkForMagic(MessageStartChars& pchMessageStart);
```
I think this is fine where it is, I like keeping these single type utilities small.

theStack reviewed (commented):
Concept ACK on the enhanced metadata, I think this includes now everything we need. Will update #27432 in a bit to the new format and test.

fjahr commented on src/node/utxo_snapshot.h:
```diff
@@ -16,17 +18,24 @@
 #include <string_view>
 
+// UTXO set snapshot magic bytes
+static constexpr uint8_t SNAPSHOT_MAGIC_BYTES[5] = {'u', 't', 'x', 'o', 0xff};
```
done

fjahr commented on src/node/utxo_snapshot.h:
```diff
@@ -16,17 +18,24 @@
     //! The hash of the block that reflects the tip of the chain for the
     //! UTXO set contained in this snapshot.
     uint256 m_base_blockhash;
+    int m_base_blockheight;
```
done

fjahr commented:
Rebased, resolved conflicts (the functional tests have changed a bit now), and addressed the open feedback. Thanks everyone for reviewing!

TheCharlatan commented on src/rpc/blockchain.cpp:
```diff
@@ -2712,7 +2712,7 @@ UniValue CreateUTXOSnapshot(
         afile << last_hash;
         WriteCompactSize(afile, coins.size());
-        for (auto [n, coin] : coins) {
+        for (const auto& [n, coin] : coins) {
```
In commit b2cf3f77b14bccf11ad5bc211db7f746ae5073fb:
Nit: I think this should be moved to the first commit.

fjahr replied:
done

TheCharlatan commented on src/kernel/chainparams.cpp:
```diff
@@ -542,3 +543,33 @@ std::unique_ptr<const CChainParams> CChainParams::TestNet()
     return std::make_unique<const CTestNetParams>();
 }
+
+std::vector<int> CChainParams::GetAvailableSnapshotHeights() const {
```
In commit b2cf3f77b14bccf11ad5bc211db7f746ae5073fb
Nit (clang-format-diff): Open braces on new line (here and the other new functions and methods below).

fjahr replied:
done

sedited reviewed (commented):
Nice, just some small comments, will review again quickly once they are addressed.

fjahr commented on src/kernel/chainparams.cpp:
```diff
@@ -542,3 +543,33 @@ std::unique_ptr<const CChainParams> CChainParams::TestNet()
+    const auto mainnet_msg = CChainParams::Main()->MessageStart();
+    const auto testnet_msg = CChainParams::TestNet()->MessageStart();
+    auto regtest_opts = CChainParams::RegTestOptions{};
+    const auto regtest_msg = CChainParams::RegTest(regtest_opts)->MessageStart();
```
done

fjahr commented on src/kernel/chainparams.cpp:
```diff
@@ -14,6 +14,7 @@
 #include <logging.h>
 #include <primitives/block.h>
 #include <primitives/transaction.h>
+#include <protocol.h>
```
Yeah, seems so, done

fjahr commented on src/node/utxo_snapshot.h:
```diff
@@ -35,11 +45,55 @@ class SnapshotMetadata
+        std::array<uint8_t, SNAPSHOT_MAGIC_BYTES.size()> snapshot_magic;
+        s >> snapshot_magic;
+        if (snapshot_magic != SNAPSHOT_MAGIC_BYTES) {
+            throw std::ios_base::failure("Invalid UTXO set snapshot magic bytes. Please check if this is indeed a snapshot file or if you are using an outdated snapshot format.", std::error_code{});
```
done

fjahr commented on src/rpc/blockchain.cpp:
```diff
@@ -2804,12 +2804,26 @@ static RPCHelpMan loadtxoutset()
+        for (auto it = available_heights.begin(); it != available_heights.end(); ++it) {
+            oss << (it != available_heights.begin() ? ", " : "") << *it;
+        }
+        std::string heights_formatted = oss.str();
```
Cool, done!

sedited reviewed (approved):
Re-ACK 542e13b2937356810bda2c41be83c3b1675e2f2f

theStack reviewed (approved):
ACK 542e13b2937356810bda2c41be83c3b1675e2f2f

fjahr commented on src/rpc/blockchain.cpp:
```diff
@@ -2695,24 +2696,48 @@ UniValue CreateUTXOSnapshot(
     afile << metadata;
 
     COutPoint key;
+    Txid last_hash;
```
Sorry, I must have overlooked that if you already mentioned it before when I changed the type. I will address it if I have to retouch.

achow101 commented:
ACK 542e13b2937356810bda2c41be83c3b1675e2f2f

fjahr commented on src/node/utxo_snapshot.h:
```diff
@@ -35,11 +45,55 @@ class SnapshotMetadata
+    inline void Serialize(Stream& s) const {
+        s << SNAPSHOT_MAGIC_BYTES;
+        s << m_version;
+        s << Params().MessageStart();
```
I addressed this in #30167 using the member field. Using params-serialization wouldn't have made much of a difference given how we use this code right now, I think.

fjahr commented on src/validation.cpp:
```diff
@@ -5735,7 +5747,8 @@ bool ChainstateManager::PopulateAndValidateSnapshot(
     bool out_of_coins{false};
     try {
-        coins_file >> outpoint;
+        Txid txid;
```
done in #30167

fjahr commented on src/validation.cpp:
```diff
@@ -5660,69 +5660,81 @@ bool ChainstateManager::PopulateAndValidateSnapshot(
+            coins_per_txid = ReadCompactSize(coins_file);
+
+            if (coins_per_txid > coins_left) {
+                LogPrintf("[snapshot] mismatch in coins count in snapshot metadata and actual snapshot data\n");
```
Added a test for this in #30167

fjahr commented:
Added release notes and dddressed post-merge comments in #30167
